13-i-can-has-extension.md

```diff
@@ -0,0 +1,191 @@
+# BOLT #13: Protocol for Permissionless Feature Development
+
+A simple feature to allow new features to be developed experimentally
+by individual implementations (and forks thereof),
+without requiring a centralized location to allocate feature bits or
+similar common resources.
+
+The intent is to allow node software to develop their own extensions
+to the protocol,
+to be later proposed to the BOLT spec once experience in live use
+has ironed out any design problems.
+
+# Table of Contents
+
+ * [Extension Encoding](#extension-encoding)
+ * [Message Type Reservation](#message-type-reservation)
+ * [Informing Extensions](#informing-extensions)
+   * [The `i_has_a_extensions` Message](#the-i_has_a_extensions-message)
+ * [Enabling Extensions](#enabling-extensions)
+   * [The `i_can_has_extension` Message](#the-i_can_has_extension-message)
+ * [Authors](#authors)
+
+# Extension Encoding
+
+Experimental features
+("extensions" in the terminology of this document)
+are identified by an extension code, a 20-byte vector of a
+white noise random distribution.
+
+An extension code SHOULD be selected
+by using RIPEMD160 on a vector generated by the concatenation of:
+
+* A human-readable UTF-8 description of the experimental feature.
+* A version number for the feature protocol.
+* A 20-byte salt from a white noise random distribution.
+
+By use of the above, developers of experimental features can
+acquire a randomly-selected extension code
+without permission from any central authority or group.
+
+An extension code MAY be selected by simply sampling twenty bytes
+from a white noise random distribution.
+
+The extension code composed of 20 `0x00` bytes is reserved to mean
+"no extension".
+
+# Message Type Reservation
+
+Message types 49152 (`0xC000`) to 65535 (`0xFFFF`) are reserved
+by this BOLT specification for extension messages.
+Two message types, `i_can_has_extension` (type 49152) and
+`i_has_a_extensions` (type 49153) are used to inform about extensions
+and enable an extension to be used over the peer protocol.
+
+Message types 49154 (`0xC002`) to 65535 (`0xFFFF`) are to be used
+for extension messages.
+These message types are _multiplexed_,
+meaning that all extensions share the same message type numbers.
+Messages for a specific extension are called _extension-specific_
+messages.
```

Comment:
How is the specific extension enabled?  (further below I see via `I_can_has_extension`) what if two peers enable extensions concurrently? Is only the recipient matching incoming messages to the previously received `I_can_has_extension` message? Shouldn't there be some form of handshake in which two peers negotiate that they will now use this particular extension? Or shouldn't there at least in message types 49154 to 65535 always be the requirement to send the current extension vector?

Reply:
Only the recipient matches incoming messages to the previously requested extension of the sender.  So one side can be talking in one extension, while the other side is talking in a different extension, and it is up to the listener to remember what "language" it is supposed to receive.  This reflects the fact that we can expect the messages to arrive at the same order (thanks to TCP) but we cannot expect both peers to execute in tight synchrony.